src/hooks/server_changed.rs

```diff
@@ -296,77 +307,113 @@ fn expand_volume(peers: Vec<Peer>, volume_info: Option<Volume>) -> Result<i32, S
+        let is_block_device = block::is_block_device(&device_path).unwrap_or(false);
+        if !is_block_device {
+            log!("Skipping invalid block device");
             continue;
         }
-        // If ephemeral-unmount is set and the directory is mounted we unmount it.
-        // Otherwise nothing happens
-        // TODO: Should this fail the hook or just skip?
-        log!("Checking for ephemeral unmount");
-        ephemeral_unmount().map_err(|e| Status::InvalidConfig(e))?;
-
+        log!(format!("Checking if {:?} is initialized", &device_path));
+        let initialized = device_initialized(&device_path).unwrap_or(false);
+        let mount_path = format!("/mnt/{}", brick_filename.to_string_lossy());
+        brick_devices.push(Device {
+            is_block_device: is_block_device,
+            // All devices start at initialized is false
+            initialized: initialized,
+            dev_path: device_path.clone(),
+            mount_path: mount_path,
+        });
+    }
+    log!("Gathering list of juju storage brick devices");
+    //Get juju storage devices
+    for brick in juju::storage_list().unwrap_or("".to_string()).lines() {
+        // This is the /dev/ location.
+        let storage_location = juju::storage_get(brick.trim()).unwrap();
         // Translate to mount location
-        let brick_filename = match brick_path.file_name() {
+        let device_path = PathBuf::from(storage_location.trim());
+        let brick_filename = match device_path.file_name() {
             Some(name) => name,
             None => {
                 log!(format!("Unable to determine filename for device: {:?}. Skipping",
-                             brick_path),
+                             brick),
                      Error);
                 continue;
             }
         };
-        // This needs to be called for the user config devices but not for juju storage devices
-        // Juju storage has a separate hook that fires to ensure the devices are already setup
-        if !device_initialized(&brick_path).map_err(|e| Status::InvalidConfig(e.to_string()))? {
-            log!(format!("Calling initialize_storage for {:?}", brick_path));
-            initialize_storage(&brick_path).map_err(|e| Status::FailedToCreate(e.to_string()))?;
-        }
-
         let mount_path = format!("/mnt/{}", brick_filename.to_string_lossy());
-        brick_paths.push(mount_path);
+        log!(format!("Checking if {:?} is initialized", &device_path));
+        let initialized = device_initialized(&device_path).unwrap_or(false);
+
+        brick_devices.push(Device {
+            is_block_device: true,
+            initialized: initialized,
+            dev_path: device_path.clone(),
+            mount_path: mount_path,
+        });
     }
+    log!(format!("storage devices: {:?}", brick_devices));
+
+    // Format all drives in parallel
+    crossbeam::scope(|scope| for device in &mut brick_devices {
```

Comment:
Rather than pass in a `&mut brick_devices`, can crossbeam give you something equivalent to a `map` so that you could just have the result of this block be a list of initialized bricks?